# Project: Evaluating and Cleaning Sales Data of a UK E-commerce Company

## Object

The goal of this data analysis is to explore and evaluate sales data to identify potenial high-performing products and assist in designing effective marketing strategies to boost revenue. 

This projectn aims to practice data cleaning and preprocessing, evalaute data quality, and prepare the dataset for further analytical work.

## Introduction

This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

**InvoiceNo**: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.

**StockCode**: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.

**Description**: Product (item) name. Nominal.

**Quantity**: The quantities of each product (item) per transaction. Numeric.

**InvoiceDate**: Invice Date and time. Numeric, the day and time when each transaction was generated.

**UnitPrice**: Unit price. Numeric, Product price per unit in sterling.

**CustomerID**: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.

**Country**: Country name. Nominal, the name of the country where each customer resides.

## Data import

To begin the analysis, the dataset is imported using read_csv function from pandas library. The file e_commerce.csv is read and converted into a Dataframe, which is assigned to the variable orginal_data.

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv(r"C:\Users\bekye\Desktop\vmlogs\Python_DA\e_commerce.csv")

In [3]:
#View the first 5 rows of the DataFrame to check the overall structure of the dataset and verify that the columns and data were imported correctly.
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## Data Evaluation

This step evaluates the dataset contained in the Dataframe orignal_data, focusing on data structure and data integrity. 
structural evaluation checks whether each column represents a variable, each row an observation, and each cell a single value.
Integrity evaluation examines issues such as missing data, duplicates,and invalid entries.

### Evaluating Data structure

A random sample of 10 rows was extracted using command:

In [4]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
114578    546028     82486   WOOD S/3 CABINET ANT WHITE FINISH         2   
151786    549568     20974         12 PENCILS SMALL TUBE SKULL        48   
368078    568938     23251        VINTAGE RED ENAMEL TRIM MUG          5   
282114    561626     22496    SET OF 2 ROUND TINS DUTCH CHEESE         1   
95530     544460    84520B  PACK 20 ENGLISH ROSE PAPER NAPKINS         3   
123363    546896     84969  BOX OF 6 ASSORTED COLOUR TEASPOONS         4   
414696    572475     21239                  PINK  POLKADOT CUP         1   
489299    578008     84992      72 SWEETHEART FAIRY CAKE CASES         1   
294913    562727     22637               PIGGY BANK RETROSPOT          4   
101558    544928     21791  VINTAGE HEADS AND TAILS CARD GAME          1   

             InvoiceDate  UnitPrice  CustomerID         Country  
114578     3/9/2011 8:33       8.95     17049.0  United Kingdom  
151786   4/10/2011 15:09       0.65     15665.0  United Kingdom  
368078   9/29/2011 14:46       1.25     17220.0  United Kingdom  
282114   7/28/2011 14:03       2.95     14898.0  United Kingdom  
95530    2/20/2011 14:01       0.85     16931.0  United Kingdom  
123363   3/17/2011 18:24       3.29         NaN  United Kingdom  
414696  10/24/2011 13:37       0.85     15271.0  United Kingdom  
489299  11/22/2011 12:47       0.55     14657.0  United Kingdom  
294913     8/9/2011 9:26       2.55     15601.0  United Kingdom  
101558   2/24/2011 18:07       2.46         NaN  United Kingdom

The sample shows that each column corresponds to a variable and each row represents a transaction.
Each cell contains a single value, including that the dataset meets the basic requirements of a tidy data structure.

### Evaluating Data Cleanliness

Assess the completeness and overall structure of the dataset.

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


The dataset contains 541909 reocrds and 8 variables.
The variable Discrption and CustomerID contain mising value.
The current data types are incorrect: InvoiceDate should be converted to datetime, and CustomerID should be converted to string.

### Evaluating Missing Data

To assess missing values, observations with  null entries int he variable Description were extracted for the inspection

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

The putput shows 1454 transaction where Description is missing. Most of these records also having a UnitPrice of 0.0, indicating that they may represent cancled or invalid transactions.

To confirm this pattern, additional filtering was applied to find records with missing Description but the Unitprice not equal to zero

In [7]:
original_data[original_data["Description"].isnull() & original_data["UnitPrice"] != 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

No records were found, confirming that all the missing Description entries correspond to transactions without valid UnitPrice values. Since Descprtion and UnitPrice are key variable for transaction analysism these row provide no meaningful information and cen be removed in later cleaning steps.

Cheking the missing value in the customerId column

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

The output shows 135080 rows with missing CustomerID. These rows still include valid transaction details such as product descirption and price, so they represent legitimate purchases made by unidentified customers.Therefore, these observations are retained for further analysis.

### Evaluating Duplicate Data

Potentail inconsistencies may exist in the country variable. Checks whether variations represent the same country

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


The output displays all distinct country names and their frequencies. Variations such as “USA” and “United States,” or “United Kingdom,” “UK,” and “U.K.” refer to the same country and should be standardized for consistency.

### Evaluating Invalid or Erroneous Data

Obtain a quick overview of the nemerical statistics to identify unusual or invalid values.

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

The result show that both Quanyity and UnitPrice contain negative values, which may affect subsequent analysis.

To futher examine these anomalies, Transaction with negative Quantity values are extracted for detailed evaluation.

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

Negative Values in the quantity variable usually indicate canceled orders, as the InvoiceNo for such transactions typically gebins with the letter "C".

To confirm this assumption, check whether any negative Quantity values wxistwith InvoiceNo not starting with "C"

In [12]:
original_data[(original_data["Quantity"] < 0 ) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

The result shows that 1336 records meet this condition, indicating that some negative Quantity values do not correspond to canceled invoices

However, most of these transactions have UnitPrice equal to zero,suggesting that they are invalid or incomplete recoreds rather than ligitimate sales.

Therefore, verify whether any negative Quantity values have non-zero UnitPrice.

In [13]:
original_data[(original_data["Quantity"] < 0 ) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []


The analysis confirmed that when Quantity is negative, the observations generally meet one of the following conditions:

- InvoiceNo begins with "C", indicating an order cancellation  
- UnitPrice = 0, representing a zero-value transaction

Such records are invalid for sales analysis and need to be removed.

Check whether UnitPrice contains negative values.

In [14]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

The output shows that all the negative UnitPrice correspond to entries labled as "Adjust bad debt". These values do not represent actual sales transaction and will be removed. 

### Data Cleaning

Based on the evaluation results, the following cleaning steps are required:
- Convert InvoiceDate from object to datetime format
- Convert CustomerId from float to string format
- Remove observations with missing Description values.
- Standardize Country values, replacing "USA" with "United stated", "UK" and "U.K." with "United Kingdom"
- Remove recoreds with negative Quantity & UnitPrice values.

To separate the cleaned dataset from the original data, create a new copy for the cleaning operations.

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

Convert InvoiceDate to the datetime format to enable time-based analysis.

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

Convert CustomerID to string format for consistent categorical representation

In [17]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

Removing the trailing .0 from CustomerId values to ensure clean and uniform formatting.

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

Remove observations with missing Description and Quantity values to ensure all records contain valid product information. After deletion, verify that no missing entries remain.

In [19]:
cleaned_data.dropna(subset = ["Description","Quantity"], inplace = True)

In [20]:
print(cleaned_data["Description"].isnull().sum())
print(cleaned_data["Quantity"].isnull().sum())

0
0


Standardize country names to maintain consistency in categorical data, and confirm that no such entries remain

In [21]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States", "UK":"United Kingdom","U.K.":"United Kingdom"})
print(len(cleaned_data[cleaned_data["Country"]=="USA"]))
print(len(cleaned_data[cleaned_data["Country"]=="UK"]))
print(len(cleaned_data[cleaned_data["Country"]=="U.K."]))

0
0
0


Remove records with negative UnitPrice values and perform the same validation check

In [22]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]
len(cleaned_data[cleaned_data["UnitPrice"]<0])

0

### Saving the Cleaned Dataset

After the data-cleaning process is completed, the cleaned dataset should be saved as a new file to ensure it can be reused in future analysis.

In [23]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [25]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [26]:
pd.read_csv("e_commerce_cleaned.csv").head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom